elasticsearch/lib/elasticsearch.rb

```diff
@@ -15,18 +15,91 @@
 # specific language governing permissions and limitations
 # under the License.
 
-require "elasticsearch/version"
-
+require 'elasticsearch/version'
 require 'elasticsearch/transport'
 require 'elasticsearch/api'
 
 module Elasticsearch
-  module Transport
-    class Client
-      include Elasticsearch::API
+  SECURITY_PRIVILEGES_VALIDATION_WARNING = 'The client is unable to verify that the server is Elasticsearch due to security privileges on the server side.'.freeze
+  NOT_ELASTICSEARCH_WARNING = 'The client noticed that the server is not Elasticsearch and we do not support this unknown product.'.freeze
+  YOU_KNOW_FOR_SEARCH = 'You know, for Search'.freeze
+
+  class Client
+    include Elasticsearch::API
+    attr_accessor :transport
+
+    # See Elasticsearch::Transport::Client for initializer parameters
+    def initialize(arguments = {}, &block)
+      @verified = false
+      @transport = Elasticsearch::Transport::Client.new(arguments, &block)
+    end
+
+    def method_missing(name, *args, &block)
+      if name == :perform_request
+        verify_elasticsearch unless @verified
+        @transport.perform_request(*args, &block)
+      else
+        super
+      end
+    end
+
+    private
+
+    def verify_elasticsearch
+      begin
+        response = elasticsearch_validation_request
+      rescue Elasticsearch::Transport::Transport::Errors::Unauthorized,
+             Elasticsearch::Transport::Transport::Errors::Forbidden
+        @verified = true
+        warn(SECURITY_PRIVILEGES_VALIDATION_WARNING)
+        return
+      end
+
+      body = if response.headers['content-type'] == 'application/yaml'
+               require 'yaml'
+               YAML.load(response.body)
+             else
+               response.body
+             end
+      version = body.dig('version', 'number')
+      verify_with_version_or_header(body, version, response.headers)
+    end
+
+    def verify_with_version_or_header(body, version, headers)
+      raise Elasticsearch::NotElasticsearchError if version.nil? || version < '6.0.0'
+
+      if version == '7.x-SNAPSHOT' || Gem::Version.new(version) >= Gem::Version.new('7.14-SNAPSHOT')
+        raise Elasticsearch::NotElasticsearchError unless headers['x-elastic-product'] == 'Elasticsearch'
+
+        @verified = true
+      elsif Gem::Version.new(version) > Gem::Version.new('6.0.0') &&
+            Gem::Version.new(version) < Gem::Version.new('7.0.0')
+        raise Elasticsearch::NotElasticsearchError unless
+          body.dig('version', 'tagline') == YOU_KNOW_FOR_SEARCH
```

Comment:
The `tagline` field is at the root of the response from ES:

```
❯ curl localhost:9200
{
  "name" : "joaos-mbp.lan",
  "cluster_name" : "elasticsearch",
  "cluster_uuid" : "CkC5-8cGSFideHtwdmnbvA",
  "version" : {
    "number" : "7.13.3",
    "build_flavor" : "default",
    "build_type" : "tar",
    "build_hash" : "5d21bea28db1e89ecc1f66311ebdec9dc3aa7d64",
    "build_date" : "2021-07-02T12:06:10.804015202Z",
    "build_snapshot" : false,
    "lucene_version" : "8.8.2",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"
}


❯ curl -s localhost:9200 | jq ".tagline"
"You Know, for Search"
```

```suggestion
          body['tagline'] == YOU_KNOW_FOR_SEARCH
```

Without this change this PR stops the client from connecting to ES 6.8

Reply:
@jsvd thanks for this fix! I wrote the wrong value on the tests so this was missed. Updating the tests with this commit :+1: